habitat-lab/habitat/datasets/rearrange/samplers/receptacle.py

```diff
@@ -241,13 +274,169 @@ def add_receptacle_visualization(
+    """
+    Defines a Receptacle surface as a triangle mesh.
+    TODO: configurable maximum height.
+    """
+
+    def __init__(
+        self,
+        name: str,
+        mesh_data: Tuple[List[Any], List[Any]],  # vertices, indices
+        parent_object_handle: str = None,
+        parent_link: Optional[int] = None,
+        up: Optional[mn.Vector3] = None,
+    ) -> None:
+        """
+        Initialize the TriangleMeshReceptacle from mesh data and pre-compute the area weighted accumulator.
+
+        :param name: The name of the Receptacle. Should be unique and descriptive for any one object.
+        :param mesh_data: The Receptacle's mesh data. A Tuple of two Lists, first vertex geometry (Vector3) and second topology (indicies of triangle corner verts(int) (len divisible by 3)).
+        :param parent_object_handle: The rigid or articulated object instance handle for the parent object to which the Receptacle is attached. None for globally defined stage Receptacles.
+        :param parent_link: Index of the link to which the Receptacle is attached if the parent is an ArticulatedObject. -1 denotes the base link. None for rigid objects and stage Receptables.
+        :param up: The "up" direction of the Receptacle in local AABB space. Used for optionally culling receptacles in un-supportive states such as inverted surfaces.
+        """
+        super().__init__(name, parent_object_handle, parent_link, up)
+        self.mesh_data = mesh_data
+        self.area_weighted_accumulator = (
+            []
+        )  # normalized float weights for each triangle for sampling
+        assert len(mesh_data[1]) % 3 == 0, "must be triangles"
+
+        # pre-compute the normalized cumulative area of all triangle faces for later sampling
+        self.total_area = 0
+        for f_ix in range(int(len(mesh_data[1]) / 3)):
+            v = self.get_face_verts(f_ix)
+            w1 = v[1] - v[0]
+            w2 = v[2] - v[1]
+            self.area_weighted_accumulator.append(
+                0.5 * np.linalg.norm(np.cross(w1, w2))
+            )
+            self.total_area += self.area_weighted_accumulator[-1]
+        for f_ix in range(len(self.area_weighted_accumulator)):
+            self.area_weighted_accumulator[f_ix] = (
+                self.area_weighted_accumulator[f_ix] / self.total_area
+            )
+            if f_ix > 0:
+                self.area_weighted_accumulator[
+                    f_ix
+                ] += self.area_weighted_accumulator[f_ix - 1]
+
+    def get_face_verts(self, f_ix: int) -> List[np.ndarray]:
+        """
+        Get all three vertices of a mesh triangle given it's face index as a list of numpy arrays.
+
+        :param f_ix: The index of the mesh triangle.
+        """
+        verts: List[np.ndarray] = []
+        for ix in range(3):
+            verts.append(
+                np.array(
+                    self.mesh_data[0][self.mesh_data[1][int(f_ix * 3 + ix)]]
```

Comment:
I don't know what is happening here. Can you create an intermediate variable for `int(f_ix * 3 + ix)` and maybe `self.mesh_data[1]`

Reply:
It's indexing a list of positions. *That* operation is clear to me, what however makes it confusing is that `self.mesh_data[0]` is a position list and `self.mesh_data[1]` is an index list. 

A named tuple instead? `self.mesh_positions` and `self.mesh_indices` instead?

Reply:
Changed to a named tuple. :+1: